### PR TITLE
Unskip and fix race in ConnectionClosedEvenIfAppSwallowsException

### PR DESCRIPTION
**Pull request by halter73**
#2464

This is the same fix @mikeharder made for RequestTimesOutWhenRequestBodyNotReceivedAtSpecifiedMinimumRate with #2589.

### CODE REVIEW

File: test/Kestrel.FunctionalTests/RequestBodyTimeoutTests.cs [hunk 1]
```diff
@@ -182,9 +189,9 @@ public class RequestBodyTimeoutTests : LoggedTest
                         "",
                         "");
 
-                    Assert.True(appRunningEvent.Wait(TestConstants.DefaultTimeout), "AppRunningEvent timed out.");
+                    await appRunningTcs.Task.DefaultTimeout();
                     systemClock.UtcNow += gracePeriod + TimeSpan.FromSeconds(1);
-                    Assert.True(exceptionSwallowedEvent.Wait(TestConstants.DefaultTimeout), "ExceptionSwallowedEvent timed out.");
+                    await exceptionSwallowedTcs.Task.DefaultTimeout();
```
Comment (ryanbrandenburg) [inline test/Kestrel.FunctionalTests/RequestBodyTimeoutTests.cs]: What are we asserting now? If there's nothing that we should/can assert I generally like to leave a comment noting the exception we're confirming we don't it, or something along those lines. That helps make it clear that we didn't just forget the Assert (which I've seen in our code-base), it was an active choice.

**Comment by ryanbrandenburg** [comment]
Just noticed that the CI failed. Looks like `Expect100ContinueHonored` is having problems where the response is empty still/again.

**Comment by halter73** [comment]
@ryanbrandenburg That's a a problem we see a lot in our macOS libuv functional tests. It's not consistent at all which test fails, but it always surfaces as an empty response.

The test ran twice with the libuv transport. You can see that in the [log for the successful test run](https://ci3.dot.net/job/aspnet_KestrelHttpServer/job/dev/job/osx-Configuration_Release_prtest/323/artifact/netci-archived-logs/Libuv.FunctionalTests/netcoreapp2.2/RequestTests/Expect100ContinueHonored-http_127.0.0.1_0.0.log), the request is handled, a response is written, and only after that does the server close the connection as indicated by the "disconnecting" and "sending FIN" logs.

```
[0.001s] [TestLifetime] [Information] Starting test "Expect100ContinueHonored-http://127.0.0.1:0" at "2018-05-31T00:28:54"
[0.002s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting starting
[0.014s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting started
[0.014s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Loaded hosting startup assembly "Libuv.FunctionalTests, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60"
[0.016s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLE6HK9PG7K6"" started.
[0.017s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request starting HTTP/1.1 POST http:///  11
[0.017s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLE6HK9PG7K6"", Request id ""0HLE6HK9PG7K6:00000001"": started reading request body.
[0.018s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLE6HK9PG7K6"", Request id ""0HLE6HK9PG7K6:00000001"": done reading request body.
[0.021s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Information] Request finished in 3.967ms 200 
[0.021s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLE6HK9PG7K6"" disconnecting.
[0.021s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv] [Debug] Connection id ""0HLE6HK9PG7K6"" sending FIN.
[0.022s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting shutdown
[0.022s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLE6HK9PG7K6"" stopped.
[0.023s] [TestLifetime] [Information] Finished test "Expect100ContinueHonored-http://127.0.0.1:0" in 0.021416s
```

In the [log for the failed test run](https://ci3.dot.net/job/aspnet_KestrelHttpServer/job/dev/job/osx-Configuration_Release_prtest/323/artifact/netci-archived-logs/Libuv.FunctionalTests/netcoreapp2.2/RequestTests/Expect100ContinueHonored-http_127.0.0.1_0.log) the client sends a FIN immediately after opening connection prior to ever sending a request, so the server appropriately closes the connection immediately. This is indicated by the "received FIN" log by the server. This can *only* be logged if a -4095 status is observed by the libuv transport's read callback. Not only is the client not supposed to send a FIN prior to sending the request (which is obvious) it's also supposed to wait to send a FIN until after it receives a complete response, so there's no way this should be happening.

```
[0.001s] [TestLifetime] [Information] Starting test "Expect100ContinueHonored-http://127.0.0.1:0" at "2018-05-31T00:28:54"
[0.008s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting starting
[0.010s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting started
[0.010s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Loaded hosting startup assembly "Libuv.FunctionalTests, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60"
[0.011s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLE6HK9PG7K5"" started.
[0.011s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv] [Debug] Connection id ""0HLE6HK9PG7K5"" received FIN.
[0.011s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLE6HK9PG7K5"" disconnecting.
[0.011s] [Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv] [Debug] Connection id ""0HLE6HK9PG7K5"" sending FIN.
[0.011s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id ""0HLE6HK9PG7K5"" stopped.
[0.016s] [Microsoft.AspNetCore.Hosting.Internal.WebHost] [Debug] Hosting shutdown
[0.018s] [TestLifetime] [Information] Finished test "Expect100ContinueHonored-http://127.0.0.1:0" in 0.016812s
```

I would say it has to be a client bug except the client in this case is Kestrel's `TestConnection` which is written directly on top of a managed Socket. The Socket isn't shutdown or disposed until the failing assertion completes.

**Comment by halter73** [comment]
@ryanbrandenburg I created a single issue to track this class of failures at #2635. If you know of any other issues you filed that relates to this behavior, it would be helpful if you could mark those as dupes of #2635 and close them.

File: test/Kestrel.FunctionalTests/RequestBodyTimeoutTests.cs (hunk 1 above)
Comment (halter73) [reply test/Kestrel.FunctionalTests/RequestBodyTimeoutTests.cs]: DefaultTimeout() is an assertion. This is common in Kestrel functional tests. Another common assertion is awaiting a TCS where SetException is set in the app func. Adding comments everywhere pointing out this is an assertion would be super noisy.

Comment (halter73) [reply test/Kestrel.FunctionalTests/RequestBodyTimeoutTests.cs]: Same goes for Receive[ForcedEnd]. That's also an assertion that we use everywhere. Anyone who works at all with Kestrel functional will be familiar with this convention.

**Comment by halter73** [comment]
Thanks for the quick approval @ryanbrandenburg

File: test/Kestrel.FunctionalTests/RequestBodyTimeoutTests.cs (hunk 1 above)
Comment (ryanbrandenburg) [reply test/Kestrel.FunctionalTests/RequestBodyTimeoutTests.cs]: Cool beans.

**Pull request merged**